conn/macos_keychain.c

```diff
@@ -0,0 +1,217 @@
+ */
+
+/**
+ * @page conn_macos_keychain Read/write macOS Keychain credentials
+ *
+ * Called by mutt_account_getuser() and mutt_user_getpass() as an alternative to
+ * storing/retrieving credentials in the configutation file, or entering them in
+ * the prompt repeatedly. A dialog will automatically pop up requesting permissions
+ * on macOS the very first time.
+ *
+ * Enabled using the configure option '--macoskeychain'.
+ */
+
+#include <Security/Security.h>
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "mutt/lib.h"
+#include "connaccount.h"
+#include "mutt_account.h"
+
+/**
+ * struct Credential - A bunch of pointers that act as a lens into ConnAccount,
+ *                     filtering out only the information relevant to keychain.
+ */
+struct Credential
+{
+  SecProtocolType protocol; ///< Connection type, e.g. HTTP
+  UInt16 port;              ///< Port to connect to
+  const char *host;         ///< Server to login to
+  const char *user;         ///< Username
+  const char *pass;         ///< Password
+};
+
+#define KEYCHAIN_ITEM(x) mutt_str_len(x), x
+
+/**
+ * find_username_in_item - Find the username in a given SecKeychainItem
+ * @param item A reference to the security keychain item
+ * @param user A pre-allocated buffer into which to write the username
+ * @retval 0 on success, a specialized error-code on failure
+ */
+static int find_username_in_item(SecKeychainItemRef item, char *user)
+{
+  SecKeychainAttributeList list = { 0 };
+  SecKeychainAttribute attr = { 0 };
+
+  list.count = 1;
+  list.attr = &attr;
+  attr.tag = kSecAccountItemAttr;
+
+  int ret = SecKeychainItemCopyContent(item, NULL, &list, NULL, NULL);
+  if (ret != 0)
+    return ret;
+
+  mutt_str_copy(user, attr.data, attr.length + 1);
+  SecKeychainItemFreeContent(&list, NULL);
+  return ret;
+}
+
+/**
+ * find_internet_password - Find a username/password pair in keychain
+ * @param user A pre-allocated buffer into which to write the username
+ * @param pass A pre-allocated buffer into which to write the password
+ * @param cred A Credential object with partial data to look up in keychain
+ * @retval 0 on success, a specialized error-code on failure
+ */
+static int find_internet_password(char *user, char *pass, struct Credential cred)
+{
+  void *buf = NULL;
+  UInt32 len = 0;
+  SecKeychainItemRef item = NULL;
+  int ret = 0;
+
+  ret = SecKeychainFindInternetPassword(NULL, /* default keychain */
+                                        KEYCHAIN_ITEM(cred.host), 0, NULL, /* account domain */
+                                        KEYCHAIN_ITEM(cred.user), 0, NULL, cred.port,
+                                        cred.protocol, kSecAuthenticationTypeDefault,
+                                        &len, &buf, &item);
+  if (ret != 0)
+    return ret;
+
+  // Keychain does not send the password with null-terminator
+  assert(len + 1 <= sizeof(pass) &&
```

Comment:
I don't understand this assertion. What's the point of sizeof(char*)?

Reply:
It's actually `char[256]` -- see the caller.

Reply:
It's not, as far as find_internet_password is concerned. You get a pointer to char.

Reply:
Hm, do you suggest changing the signature of `find_internet_password`?

Reply:
I suggest to pass the length as a parameter

Reply:
find_internet_password(account->user, sizeof(account->user), account->pass, sizeof(account->pass), cred);